File: packages/svmc-js/src/allocation/alloc-hypothesis-2.ts
/**
 * Daily biomass allocation (alloc_hypothesis_2).
 *
 * Fortran reference: vendor/SVMC/src/allocation.f90 L141–280
 *
 * Allocates daily GPP into leaf, stem, root, and grain carbon pools with
 * Q10-dependent maintenance and growth respiration, turnover-based litter
 * production, and management actions (harvest, grazing, organic fertilizer).
 *
 * All branches use np.where for differentiability through the complete
 * allocation logic.
 */

import { np } from "../precision.js";

const SPD = 3600.0 * 24.0; // seconds per day
const EPS = 1e-30;         // denominator guard

/** Guard the undefined x/0 case with a tiny sign-preserving fallback. */
function safeDenom(x: np.Array): np.Array {
  using absX = np.abs(x);
  using tooSmall = absX.less(EPS);
  using signX = np.sign(x);
  using fallback = signX.mul(EPS);
  return np.where(tooSmall, fallback, x);
}

/**
 * Compute daily biomass allocation from GPP to carbon pools.
 *
 * All operations stay on traced np.Array values so the full computation is
 * differentiable via jax-js-nonconsuming autodiff (jit/grad/vjp).
 *
 * @param tempDay - Daily mean temperature (°C).
 * @param gppDay - Gross primary production (kg C m⁻² s⁻¹).
 * @param leafRdarkDay - Leaf dark respiration (kg C m⁻² s⁻¹).
 * @param croot - Root carbon (kg C m⁻²).
 * @param cleaf - Leaf carbon (kg C m⁻²).
 * @param cstem - Stem carbon (kg C m⁻²).
 * @param cgrain - Grain carbon (kg C m⁻²).
 * @param litterCleafIn - Input leaf litter carbon (used when invert_option != 0).
 * @param grainFill - Grain-filling flux (kg C m⁻² s⁻¹).
 * @param cratioResp - Maintenance respiration fraction.
 * @param cratioLeaf - Leaf allocation fraction.
 * @param cratioRoot - Root allocation fraction.
 * @param cratioBiomass - Carbon fraction of biomass.
 * @param turnoverCleaf - Leaf turnover rate at 20°C.
 * @param turnoverCroot - Root/stem turnover rate at 20°C.
 * @param sla - Specific leaf area (m² kg⁻¹).
 * @param q10 - Q10 temperature coefficient.
 * @param invertOption - Inversion mode (0=none, 1=cratio_leaf, 2=turnover).
 * @param managementType - Management type (0=none, 1=harvest, 3=grazing, 4=organic).
 * @param managementCInput - Carbon input rate from management (kg C m⁻² s⁻¹).
 * @param managementCOutput - Carbon output rate from management (kg C m⁻² s⁻¹).
 * @param pftIsOat - 1.0 for oat, 0.0 for grass.
 * @param phenoStage - Phenological stage (1=growth, 2=dormancy).
 * @returns Object with all pool states and fluxes. Caller must dispose all returned arrays.
 */
export function allocHypothesis2Fn(
  tempDay: np.Array,
  gppDay: np.Array,
  leafRdarkDay: np.Array,
  croot: np.Array,
  cleaf: np.Array,
  cstem: np.Array,
  cgrain: np.Array,
  litterCleafIn: np.Array,
  grainFill: np.Array,
  cratioResp: np.Array,
  cratioLeaf: np.Array,
  cratioRoot: np.Array,
  cratioBiomass: np.Array,
  turnoverCleaf: np.Array,
  turnoverCroot: np.Array,
  sla: np.Array,
  q10: np.Array,
  invertOption: np.Array,
  managementType: np.Array,
  managementCInput: np.Array,
  managementCOutput: np.Array,
  pftIsOat: np.Array,
  phenoStage: np.Array,
): {
  nppDay: np.Array;
  autoResp: np.Array;
  croot: np.Array;
  cleaf: np.Array;
  cstem: np.Array;
  cgrain: np.Array;
  litterCleaf: np.Array;
  litterCroot: np.Array;
  compost: np.Array;
  lai: np.Array;
  abovebiomass: np.Array;
  belowbiomass: np.Array;
  yield: np.Array;
  grainFill: np.Array;
  phenoStage: np.Array;
} {
  using isGrowth = phenoStage.equal(1);
  using invIs0 = invertOption.equal(0.0);

  // ---------- GROWTH PHASE ----------

  // Q10 temperature factor
  using _tDiff = tempDay.sub(20.0);
  using _tRatio = _tDiff.div(10.0);
  using q10f = np.power(q10, _tRatio);

  // Growth respirations
  using _gppLeaf = gppDay.mul(cratioLeaf);
  using _gppLeafMinusRdark = _gppLeaf.sub(leafRdarkDay);
  using _gppLeafMinusRdarkSpd = _gppLeafMinusRdark.mul(SPD);
  using _grRespLeafPre = np.maximum(_gppLeafMinusRdarkSpd, 0.0);
  using grRespLeaf = _grRespLeafPre.mul(0.11);

  using _oneMinusCLCR = cratioLeaf.add(cratioRoot);
  using _oneVal = np.array(1.0);
  using _stemFrac = _oneVal.sub(_oneMinusCLCR);
  using _gppStem = gppDay.mul(_stemFrac);
  using _cstemResp1b = cstem.mul(cratioResp);
  using _cstemResp = _cstemResp1b.mul(q10f);
  using _stemNet = _gppStem.sub(_cstemResp);
  using _stemNetSpd = _stemNet.mul(SPD);
  using _grRespStemPre = np.maximum(_stemNetSpd, 0.0);
  using grRespStem = _grRespStemPre.mul(0.11);

  using _gppRoot = gppDay.mul(cratioRoot);
  using _gppRootMinusGF = _gppRoot.sub(grainFill);
  using _crootResp1b = croot.mul(cratioResp);
  using _crootResp = _crootResp1b.mul(q10f);
  using _rootNet = _gppRootMinusGF.sub(_crootResp);
  using _rootNetSpd = _rootNet.mul(SPD);
  using _grRespRootPre = np.maximum(_rootNetSpd, 0.0);
  using grRespRoot = _grRespRootPre.mul(0.11);

  using _cg1 = cgrain.mul(0.1);
  using _cg2 = _cg1.mul(cratioResp);
  using _cgrainResp = _cg2.mul(q10f);
  using _grainNet = grainFill.sub(_cgrainResp);
  using _grainNetSpd = _grainNet.mul(SPD);
  using _grRespGrainPre = np.maximum(_grainNetSpd, 0.0);
  using _grRespGrainOat = _grRespGrainPre.mul(0.11);
  using _isOat = pftIsOat.greater(0.5);
  using grRespGrain = np.where(_isOat, _grRespGrainOat, 0.0);

  // NPP and autotrophic respiration
  using _crootPlusCstem = croot.add(cstem);
  using _maintSC1 = _crootPlusCstem.mul(cratioResp);
  using maintSC = _maintSC1.mul(q10f);
  using _maintG1 = cgrain.mul(0.1);
  using _maintG2 = _maintG1.mul(cratioResp);
  using maintG = _maintG2.mul(q10f);
  using _totalMaint1 = maintSC.add(maintG);
  using _totalMaint = _totalMaint1.add(leafRdarkDay);
  using _nppPre = gppDay.sub(_totalMaint);
  using nppG = _nppPre.mul(SPD);
  using _totalMaintSpd = _totalMaint.mul(SPD);
  using _grRespLS = grRespLeaf.add(grRespStem);
  using _grRespLSR = _grRespLS.add(grRespRoot);
  using _grRespTotal = _grRespLSR.add(grRespGrain);
  using autoRespG = _totalMaintSpd.add(_grRespTotal);

  // Turnover-based litter
  using _lcTurn1 = cleaf.mul(turnoverCleaf);
  using litterCleafTurnover = _lcTurn1.mul(q10f);
  using litterCleafG = np.where(invIs0, litterCleafTurnover, litterCleafIn);
  using _lcStem1 = cstem.mul(turnoverCroot);
  using litterCstemG = _lcStem1.mul(q10f);
  using _lcRoot1 = croot.mul(turnoverCroot);
  using litterCrootG = _lcRoot1.mul(q10f);
  using compostG = np.array(0.0);

  // Update pools (before management)
  using _cleafGpp1 = gppDay.mul(cratioLeaf);
  using _cleafGppTerm = _cleafGpp1.mul(SPD);
  using _cleafBase = cleaf.add(_cleafGppTerm);
  using _cleafSub1 = _cleafBase.sub(litterCleafG);
  using _rdarkSpd = leafRdarkDay.mul(SPD);
  using _cleafSub2 = _cleafSub1.sub(_rdarkSpd);
  using _cleafInv0 = _cleafSub2.sub(grRespLeaf);
  using cleafG = np.where(invIs0, _cleafInv0, cleaf);

  using _cstemGpp1 = gppDay.mul(_stemFrac);
  using _cstemGppTerm = _cstemGpp1.mul(SPD);
  using _cstemResp1 = cstem.mul(cratioResp);
  using _cstemResp2 = _cstemResp1.mul(q10f);
  using _cstemRespSpd = _cstemResp2.mul(SPD);
  using _cstem1 = cstem.add(_cstemGppTerm);
  using _cstem2 = _cstem1.sub(_cstemRespSpd);
  using _cstem3 = _cstem2.sub(litterCstemG);
  using cstemG = _cstem3.sub(grRespStem);

  using _crootGppTerm = _gppRootMinusGF.mul(SPD);
  using _crootResp1 = croot.mul(cratioResp);
  using _crootResp2 = _crootResp1.mul(q10f);
  using _crootRespSpd = _crootResp2.mul(SPD);
  using _croot1 = croot.add(_crootGppTerm);
  using _croot2 = _croot1.sub(litterCrootG);
  using _croot3 = _croot2.sub(_crootRespSpd);
  using _crootPre = _croot3.sub(grRespRoot);
  using crootG = np.maximum(_crootPre, 0.0);

  using _cgrainGFSpd = grainFill.mul(SPD);
  using _cg01 = cgrain.mul(0.1);
  using _cg02 = _cg01.mul(cratioResp);
  using _cg03 = _cg02.mul(q10f);
  using _cgrainRespSpd = _cg03.mul(SPD);
  using _cgrainOat1 = cgrain.add(_cgrainGFSpd);
  using _cgrainOat2 = _cgrainOat1.sub(_cgrainRespSpd);
  using _cgrainOat = _cgrainOat2.sub(grRespGrain);
  using cgrainG = np.where(_isOat, _cgrainOat, 0.0);

  // --- Management ---
  using isHarvest = managementType.equal(1);
  using isGrazing = managementType.equal(3);
  using isOrganic = managementType.equal(4);

  // Harvest grass: proportional deduction from cleaf (if inv==0) and cstem
  using hgDeduct = managementCOutput.mul(SPD);
  using _hgCleafCstemSum = cleafG.add(cstemG);
  using _hgSafeDenom1 = safeDenom(_hgCleafCstemSum);
  using _hgCleafMul = hgDeduct.mul(cleafG);
  using _hgCleafFrac = _hgCleafMul.div(_hgSafeDenom1);
  using _hgCleafReduced = cleafG.sub(_hgCleafFrac);
  using hgCleaf = np.where(invIs0, _hgCleafReduced, cleafG);
  // cstem uses possibly-modified cleaf in denominator (sequential effect)
  using _hgCleafCstem2 = hgCleaf.add(cstemG);
  using _hgSafeDenom2 = safeDenom(_hgCleafCstem2);
  using _hgCstemMul = hgDeduct.mul(cstemG);
  using _hgCstemFrac = _hgCstemMul.div(_hgSafeDenom2);
  using hgCstem = cstemG.sub(_hgCstemFrac);

  // Harvest oat: overwrite litter with updated pools, zero everything
  // (ho_litter_cleaf = cleafG, ho_litter_croot = crootG, ho_litter_cstem = cstemG)

  // Harvest combined (select by pft)
  using hCleaf = np.where(_isOat, 0.0, hgCleaf);
  using hCstem = np.where(_isOat, 0.0, hgCstem);
  using hCroot = np.where(_isOat, 0.0, crootG);
  using hCgrain = np.where(_isOat, 0.0, cgrainG);
  using hNpp = np.where(_isOat, 0.0, nppG);
  using hResp = np.where(_isOat, 0.0, autoRespG);
  using hLitterCleaf = np.where(_isOat, cleafG, litterCleafG);
  using hLitterCroot = np.where(_isOat, crootG, litterCrootG);
  using hLitterCstem = np.where(_isOat, cstemG, litterCstemG);
  using hCompost = compostG;

  // Grazing: deduct from cleaf (if inv==0) and cstem, add compost
  using gzDeduct = managementCOutput.mul(SPD);
  using _gzCleafCstemSum = cleafG.add(cstemG);
  using _gzSafeDenom1 = safeDenom(_gzCleafCstemSum);
  using _gzCleafMul = gzDeduct.mul(cleafG);
  using _gzCleafFrac = _gzCleafMul.div(_gzSafeDenom1);
  using _gzCleafReduced = cleafG.sub(_gzCleafFrac);
  using gzCleaf = np.where(invIs0, _gzCleafReduced, cleafG);
  using _gzCleafCstem2 = gzCleaf.add(cstemG);
  using _gzSafeDenom2 = safeDenom(_gzCleafCstem2);
  using _gzCstemMul = gzDeduct.mul(cstemG);
  using _gzCstemFrac = _gzCstemMul.div(_gzSafeDenom2);
  using gzCstem = cstemG.sub(_gzCstemFrac);
  using gzCompost = managementCInput.mul(SPD);

  // Organic: only compost
  using orgCompost = managementCInput.mul(SPD);

  // Select by management type
  function sel(h: np.Array, gz: np.Array, org: np.Array, nm: np.Array): np.Array {
    using _inner1 = np.where(isOrganic, org, nm);
    using _inner2 = np.where(isGrazing, gz, _inner1);
    return np.where(isHarvest, h, _inner2);
  }

  using mCleaf = sel(hCleaf, gzCleaf, cleafG, cleafG);
  using mCstem = sel(hCstem, gzCstem, cstemG, cstemG);
  using mCroot = sel(hCroot, crootG, crootG, crootG);
  using mCgrain = sel(hCgrain, cgrainG, cgrainG, cgrainG);
  using mNpp = sel(hNpp, nppG, nppG, nppG);
  using mResp = sel(hResp, autoRespG, autoRespG, autoRespG);
  using mLitterCleaf = sel(hLitterCleaf, litterCleafG, litterCleafG, litterCleafG);
  using mLitterCroot = sel(hLitterCroot, litterCrootG, litterCrootG, litterCrootG);
  using mLitterCstem = sel(hLitterCstem, litterCstemG, litterCstemG, litterCstemG);
  using mCompost = sel(hCompost, gzCompost, orgCompost, compostG);

  // Combine leaf + stem litter
  using gLitterCleaf = mLitterCleaf.add(mLitterCstem);
  using gLitterCroot = mLitterCroot;

  // Derived quantities (growth phase)
  using _aboveLS = mCleaf.add(mCstem);
  using _gAboveSum = _aboveLS.add(mCgrain);
  using gAbove = _gAboveSum.div(cratioBiomass);
  using gBelow = mCroot.div(cratioBiomass);
  using _gLaiPre = mCleaf.div(cratioBiomass);
  using gLai = _gLaiPre.mul(sla);

  // ---------- DORMANCY PHASE (pheno_stage == 2) ----------
  // Original input values (NOT combined with cstem)
  using dLitterCleaf = cleaf;
  using dLitterCroot = croot;
  using dAbove = cgrain.div(cratioBiomass);
  using dBelow = np.array(0.0);
  using dLai = np.array(0.0);

  // ---------- SELECT BY PHENOLOGICAL STAGE ----------
  return {
    nppDay: np.where(isGrowth, mNpp, 0.0),
    autoResp: np.where(isGrowth, mResp, 0.0),
    croot: np.where(isGrowth, mCroot, 0.0),
    cleaf: np.where(isGrowth, mCleaf, 0.0),
    cstem: np.where(isGrowth, mCstem, 0.0),
    cgrain: np.where(isGrowth, mCgrain, cgrain),
    litterCleaf: np.where(isGrowth, gLitterCleaf, dLitterCleaf),
    litterCroot: np.where(isGrowth, gLitterCroot, dLitterCroot),
    compost: np.where(isGrowth, mCompost, 0.0),
    lai: np.where(isGrowth, gLai, dLai),
    abovebiomass: np.where(isGrowth, gAbove, dAbove),
    belowbiomass: np.where(isGrowth, gBelow, dBelow),
    yield: np.array(0.0),    // never modified
    grainFill,               // never modified — caller owns lifetime
    phenoStage: np.where(isGrowth, 1, 1),  // both paths reset to 1
  };
}

File: packages/svmc-jax/src/svmc_jax/allocation/alloc_hypothesis_2.py
"""Daily biomass allocation (alloc_hypothesis_2).

Fortran reference: vendor/SVMC/src/allocation.f90 L141–280

Allocates daily GPP into leaf, stem, root, and grain carbon pools with
Q10-dependent maintenance and growth respiration, turnover-based litter
production, and management actions (harvest, grazing, organic fertilizer).

The implementation uses jnp.where for all branches to maintain JAX
differentiability through the complete allocation logic.
"""

import functools
from typing import NamedTuple

import jax
import jax.numpy as jnp

_SPD = 3600.0 * 24.0  # seconds per day


class AllocHypothesisResult(NamedTuple):
    """Typed allocation result container with backward-compatible key access."""

    npp_day: jnp.ndarray
    auto_resp: jnp.ndarray
    croot: jnp.ndarray
    cleaf: jnp.ndarray
    cstem: jnp.ndarray
    cgrain: jnp.ndarray
    litter_cleaf: jnp.ndarray
    litter_croot: jnp.ndarray
    compost: jnp.ndarray
    lai: jnp.ndarray
    abovebiomass: jnp.ndarray
    belowbiomass: jnp.ndarray
    yield_: jnp.ndarray
    grain_fill: jnp.ndarray
    pheno_stage: jnp.ndarray

    def __getitem__(self, item):
        if item == "yield":
            return self.yield_
        if isinstance(item, str):
            return getattr(self, item)
        return tuple.__getitem__(self, item)


@functools.partial(jax.jit, static_argnames=())
def alloc_hypothesis_2(
    temp_day, gpp_day, leaf_rdark_day,
    croot, cleaf, cstem, cgrain,
    litter_cleaf_in, grain_fill,
    cratio_resp, cratio_leaf, cratio_root, cratio_biomass,
    turnover_cleaf, turnover_croot, sla, q10, invert_option,
    management_type, management_c_input, management_c_output,
    pft_is_oat, pheno_stage,
):
    """Compute daily biomass allocation from GPP to carbon pools.

    Fortran reference: subroutine alloc_hypothesis_2 in allocation.f90

    Args:
        temp_day: Daily mean temperature (°C).
        gpp_day: Gross primary production (kg C m⁻² s⁻¹).
        leaf_rdark_day: Leaf dark respiration (kg C m⁻² s⁻¹).
        croot: Root carbon (kg C m⁻²).
        cleaf: Leaf carbon (kg C m⁻²).
        cstem: Stem carbon (kg C m⁻²).
        cgrain: Grain carbon (kg C m⁻²).
        litter_cleaf_in: Input leaf litter carbon (used when invert_option != 0).
        grain_fill: Grain-filling flux (kg C m⁻² s⁻¹).
        cratio_resp: Maintenance respiration fraction.
        cratio_leaf: Leaf allocation fraction.
        cratio_root: Root allocation fraction.
        cratio_biomass: Carbon fraction of biomass.
        turnover_cleaf: Leaf turnover rate at 20°C.
        turnover_croot: Root/stem turnover rate at 20°C.
        sla: Specific leaf area (m² kg⁻¹).
        q10: Q10 temperature coefficient.
        invert_option: Inversion mode (0=none, 1=cratio_leaf, 2=turnover).
        management_type: Management type (0=none, 1=harvest, 3=grazing, 4=organic).
        management_c_input: Carbon input rate from management (kg C m⁻² s⁻¹).
        management_c_output: Carbon output rate from management (kg C m⁻² s⁻¹).
        pft_is_oat: 1.0 for oat, 0.0 for grass.
        pheno_stage: Phenological stage (1=growth, 2=dormancy).

    Returns:
        AllocHypothesisResult with fields: npp_day, auto_resp, croot, cleaf,
        cstem, cgrain, litter_cleaf, litter_croot, compost, lai,
        abovebiomass, belowbiomass, yield_, grain_fill, pheno_stage.
    """
    spd = _SPD
    is_growth = pheno_stage == 1
    inv_is_0 = invert_option == 0.0

    # ---------- GROWTH PHASE (pheno_stage == 1) ----------

    # Q10 temperature factor
    q10f = q10 ** ((temp_day - 20.0) / 10.0)

    # Growth respirations
    gr_resp_leaf = jnp.maximum(
        0.0, (gpp_day * cratio_leaf - leaf_rdark_day) * spd) * 0.11
    gr_resp_stem = jnp.maximum(
        (gpp_day * (1.0 - cratio_leaf - cratio_root)
         - cstem * cratio_resp * q10f) * spd, 0.0) * 0.11
    gr_resp_root = jnp.maximum(
        (gpp_day * cratio_root - grain_fill
         - croot * cratio_resp * q10f) * spd, 0.0) * 0.11
    gr_resp_grain = jnp.where(
        pft_is_oat,
        jnp.maximum(
            (grain_fill - cgrain * 0.1 * cratio_resp * q10f) * spd,
            0.0) * 0.11,
        0.0)

    # NPP and autotrophic respiration
    maint_sc = (croot + cstem) * cratio_resp * q10f
    maint_g = cgrain * 0.1 * cratio_resp * q10f
    npp_g = (gpp_day - maint_sc - maint_g - leaf_rdark_day) * spd
    auto_resp_g = (maint_sc + maint_g + leaf_rdark_day) * spd \
        + gr_resp_leaf + gr_resp_stem + gr_resp_root + gr_resp_grain

    # Turnover-based litter
    litter_cleaf_turnover = cleaf * turnover_cleaf * q10f
    litter_cleaf_g = jnp.where(inv_is_0, litter_cleaf_turnover, litter_cleaf_in)
    litter_cstem_g = cstem * turnover_croot * q10f
    litter_croot_g = croot * turnover_croot * q10f
    compost_g = jnp.float64(0.0)

    # Update pools (before management)
    cleaf_g = jnp.where(
        inv_is_0,
        cleaf + gpp_day * cratio_leaf * spd
        - litter_cleaf_g - leaf_rdark_day * spd - gr_resp_leaf,
        cleaf)
    cstem_g = (cstem + gpp_day * (1.0 - cratio_leaf - cratio_root) * spd
               - cstem * cratio_resp * q10f * spd
               - litter_cstem_g - gr_resp_stem)
    croot_g = jnp.maximum(
        0.0,
        croot + (gpp_day * cratio_root - grain_fill) * spd - litter_croot_g
        - croot * cratio_resp * q10f * spd - gr_resp_root)
    cgrain_g = jnp.where(
        pft_is_oat,
        cgrain + grain_fill * spd
        - cgrain * 0.1 * cratio_resp * q10f * spd - gr_resp_grain,
        0.0)

    # --- Management ---
    def _safe_denom(x):
        """Guard the undefined x/0 case with a tiny sign-preserving fallback."""
        return jnp.where(jnp.abs(x) < 1e-30, jnp.copysign(1e-30, x), x)

    # Harvest grass: proportional deduction from cleaf (if inv==0) and cstem
    hg_deduct = management_c_output * spd
    hg_cleaf_frac = hg_deduct * cleaf_g / _safe_denom(cleaf_g + cstem_g)
    hg_cleaf = jnp.where(inv_is_0, cleaf_g - hg_cleaf_frac, cleaf_g)
    # cstem uses possibly-modified cleaf in denominator (sequential effect)
    hg_cstem = cstem_g - hg_deduct * cstem_g / _safe_denom(hg_cleaf + cstem_g)

    # Harvest oat: overwrite litter with updated pools, zero everything
    ho_litter_cleaf = cleaf_g
    ho_litter_croot = croot_g
    ho_litter_cstem = cstem_g

    # Harvest combined (select by pft)
    h_cleaf = jnp.where(pft_is_oat, 0.0, hg_cleaf)
    h_cstem = jnp.where(pft_is_oat, 0.0, hg_cstem)
    h_croot = jnp.where(pft_is_oat, 0.0, croot_g)
    h_cgrain = jnp.where(pft_is_oat, 0.0, cgrain_g)
    h_npp = jnp.where(pft_is_oat, 0.0, npp_g)
    h_resp = jnp.where(pft_is_oat, 0.0, auto_resp_g)
    h_litter_cleaf = jnp.where(pft_is_oat, ho_litter_cleaf, litter_cleaf_g)
    h_litter_croot = jnp.where(pft_is_oat, ho_litter_croot, litter_croot_g)
    h_litter_cstem = jnp.where(pft_is_oat, ho_litter_cstem, litter_cstem_g)
    h_compost = compost_g

    # Grazing: deduct from cleaf (if inv==0) and cstem, add compost
    gz_deduct = management_c_output * spd
    gz_cleaf_frac = gz_deduct * cleaf_g / _safe_denom(cleaf_g + cstem_g)
    gz_cleaf = jnp.where(inv_is_0, cleaf_g - gz_cleaf_frac, cleaf_g)
    gz_cstem = cstem_g - gz_deduct * cstem_g / _safe_denom(gz_cleaf + cstem_g)
    gz_compost = management_c_input * spd

    # Organic: only compost
    org_compost = management_c_input * spd

    # No management: pass through
    is_harvest = management_type == 1
    is_grazing = management_type == 3
    is_organic = management_type == 4

    def _sel(h, gz, org, nm):
        """Select value by management type."""
        return jnp.where(is_harvest, h,
               jnp.where(is_grazing, gz,
               jnp.where(is_organic, org, nm)))

    m_cleaf = _sel(h_cleaf, gz_cleaf, cleaf_g, cleaf_g)
    m_cstem = _sel(h_cstem, gz_cstem, cstem_g, cstem_g)
    m_croot = _sel(h_croot, croot_g, croot_g, croot_g)
    m_cgrain = _sel(h_cgrain, cgrain_g, cgrain_g, cgrain_g)
    m_npp = _sel(h_npp, npp_g, npp_g, npp_g)
    m_resp = _sel(h_resp, auto_resp_g, auto_resp_g, auto_resp_g)
    m_litter_cleaf = _sel(h_litter_cleaf, litter_cleaf_g, litter_cleaf_g, litter_cleaf_g)
    m_litter_croot = _sel(h_litter_croot, litter_croot_g, litter_croot_g, litter_croot_g)
    m_litter_cstem = _sel(h_litter_cstem, litter_cstem_g, litter_cstem_g, litter_cstem_g)
    m_compost = _sel(h_compost, gz_compost, org_compost, compost_g)

    # Combine leaf + stem litter
    g_litter_cleaf = m_litter_cleaf + m_litter_cstem
    g_litter_croot = m_litter_croot

    # Derived quantities (growth phase)
    g_above = (m_cleaf + m_cstem + m_cgrain) / cratio_biomass
    g_below = m_croot / cratio_biomass
    g_lai = m_cleaf / cratio_biomass * sla
    g_pheno = 1  # unchanged in growth

    # ---------- DORMANCY PHASE (pheno_stage == 2) ----------
    d_litter_cleaf = cleaf   # original input value (NOT combined with cstem)
    d_litter_croot = croot   # original input value
    d_above = cgrain / cratio_biomass  # only cgrain survives (not zeroed)
    d_below = jnp.float64(0.0)
    d_lai = jnp.float64(0.0)
    d_pheno = 1  # reset to growth

    # ---------- SELECT BY PHENOLOGICAL STAGE ----------
    return AllocHypothesisResult(
        npp_day=jnp.where(is_growth, m_npp, 0.0),
        auto_resp=jnp.where(is_growth, m_resp, 0.0),
        croot=jnp.where(is_growth, m_croot, 0.0),
        cleaf=jnp.where(is_growth, m_cleaf, 0.0),
        cstem=jnp.where(is_growth, m_cstem, 0.0),
        cgrain=jnp.where(is_growth, m_cgrain, cgrain),
        litter_cleaf=jnp.where(is_growth, g_litter_cleaf, d_litter_cleaf),
        litter_croot=jnp.where(is_growth, g_litter_croot, d_litter_croot),
        compost=jnp.where(is_growth, m_compost, 0.0),
        lai=jnp.where(is_growth, g_lai, d_lai),
        abovebiomass=jnp.where(is_growth, g_above, d_above),
        belowbiomass=jnp.where(is_growth, g_below, d_below),
        yield_=jnp.float64(0.0),  # never modified
        grain_fill=grain_fill,    # never modified
        pheno_stage=jnp.where(is_growth, g_pheno, d_pheno),
    )
